rfd/0107-desktop-access-performance-research.md

```diff
@@ -0,0 +1,118 @@
+```
+
+- server sends [message](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/3c76f8fb-fcaa-4ca6-9d2e-3f449bdbed40) to inform client to store bitmap in the cache
+- windows destkop service translates rdp message into tdp message to store bitmap and sends it to the proxy
+- proxy forwards message to web browser
+- web browser stores bitmap in the in-memory cache
+- server sends [message](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/84c2ec2f-f776-405b-9b48-6894a28b1b14) to inform client to load bitmap from the cache
+- windows destkop service translates rdp message into tdp message to load bitmap and sends it to the proxy
+- proxy forwards message to web browser
+- web browser loads bitmap from the in-memory cache
+
+### Process bitmaps in Rust library
+While interacting with the remote desktop using the RDP protocol, most of the protocol messages exchanged between the server and a client are related to rendering bitmaps. Messages almost always contain compressed data to reduce bandwidth usage and latency. Rendering compressed bitmaps on the screen requires uncompressing data first using decompress algorithm and encoding bitmaps into PNG.
+While decompression is already done in the Rust library, the encoding of bitmaps is done in the Go client.
+
+The best performance can be achieved by moving the PNG encoding procedure from the Go client to the Rust library. It'll also simplify the way we encode and decode PNG TDP messages.
+During the tests, the average time it took to process messages (read, process, decompress, and encode) when the encoding took place in the Go side was around 500μs. After moving the encoding of bitmaps into PNGs to the Rust library, the time it took to process a message went down to 50μs.
+
+
+
+### Remote Desktop Protocol: Graphics Pipeline Extension  [MS-RDPEGFX](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegfx/da5c75f9-cd99-450c-98c4-014a496942b0)
+
+
+The [MS-RDPEGFX] extension introduces a new way of drawing data by creating surfaces and using commands that utilize these surfaces to draw frames rather than just rendering bitmap data on the screen. This extension efficiently encode and transmit graphic display data from the server to the client. It alsow allows caching data locally to improve performance, can use different, newer, and more performant codecs to encode data e.g:
+- RemoteFX
+- ClearCodec
+- MPEG-4 AVC/H.264
+
+
+In essence, this protocol works in this way:
+- create surfaces
+- associate surfaces to the portions of graphics output buffer (this is what an user sees) that can be of different sizes
+- decode graphic data for surface and render it on the graphics output buffer
+
+There is already one implementation of the extension in the Rust language: [ironrdp-gui-client](https://github.com/Devolutions/IronRDP/tree/master/ironrdp-gui-client), but it is using native gui for rendering by utilising opengl. 
```

Comment:
What makes you think this is using `MS-RDPEGFX`? From what I see, it's using the same `process_connection_sequence` and therefore the same capabilities exchange as other clients in the repo, ergo it's using the same version of the protocol.

That said, [apparently some of GFX was implemented for video streaming](https://github.com/Devolutions/IronRDP/issues/76#issuecomment-1433676943).

Reply:
Yes that's right that it's using the `process_connection_sequence` but you can pass your own capabilites or [enable the avc444 / h264 in cofing](https://github.com/Devolutions/IronRDP/blob/master/ironrdp-gui-client/src/config.rs#L120) and then that config is passed to the active session processor.

and then:
- [start gui](https://github.com/Devolutions/IronRDP/blob/master/ironrdp-gui-client/src/main.rs#L153)
- [create renderer and pass window to it](https://github.com/Devolutions/IronRDP/blob/master/ironrdp-gui-client/src/gui.rs#L82)
- and then this render starts two threads, one for [processing [MS-RDPGFX] messages](https://github.com/Devolutions/IronRDP/blob/master/ironrdp-renderer/src/renderer.rs#L147) and [another for drawing](https://github.com/Devolutions/IronRDP/blob/master/ironrdp-renderer/src/renderer.rs#L152)

Anyway if you don't enable avc444 / h264 it'll use the bitmaps encoded in the RemoteFX. Another thing is that this gui client doesn't work on macOS without a lot of fixes

Reply:
I see, thanks for linking me to it